src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,337 @@
+///
+/// * any pointer to a memory block which is [*currently allocated*] may be passed to any other
+///   method of the allocator.
+///
+/// [*currently allocated*]: #currently-allocated-memory
+#[unstable(feature = "allocator_api", issue = "32838")]
+pub unsafe trait AllocRef {
+    /// On success, returns a pointer meeting the size and alignment guarantees of `layout` and the
+    /// actual size of the allocated block, which is greater than or equal to `layout.size()`.
+    ///
+    /// If this method returns an `Ok((addr, actual_size))`, then the `addr` returned will be
+    /// non-null address pointing to a block of storage suitable for holding an instance represented
+    /// by `layout`.
+    ///
+    /// The returned block of storage is initialized like specified by [`init`], all the way up to
+    /// the returned `actual_size`.
+    ///
+    /// [`init`]: AllocInit
+    ///
+    /// # Errors
+    ///
+    /// Returning `Err` indicates that either memory is exhausted or `layout` does not meet
+    /// allocator's size or alignment constraints.
+    ///
+    /// Implementations are encouraged to return `Err` on memory exhaustion rather than panicking or
+    /// aborting, but this is not a strict requirement. (Specifically: it is *legal* to implement
+    /// this trait atop an underlying native allocation library that aborts on memory exhaustion.)
+    ///
+    /// Clients wishing to abort computation in response to an allocation error are encouraged to
+    /// call the [`handle_alloc_error`] function, rather than directly invoking `panic!` or similar.
+    ///
+    /// [`handle_alloc_error`]: ../../alloc/alloc/fn.handle_alloc_error.html
+    fn alloc(&mut self, layout: Layout, init: AllocInit) -> Result<(NonNull<u8>, usize), AllocErr>;
+
+    /// Deallocate the memory referenced by `ptr`.
+    ///
+    /// # Safety
+    ///
+    /// * `ptr` must denote a block of memory [*currently allocated*] via this allocator,
+    ///
+    /// * `layout` must [*fit*] that block of memory, and
+    ///
+    /// * the alignment of the `layout` must match the alignment used to allocate that block of
+    ///   memory.
+    ///
+    /// [*currently allocated*]: #currently-allocated-memory
+    /// [*fit*]: #memory-fitting
+    unsafe fn dealloc(&mut self, ptr: NonNull<u8>, layout: Layout);
+
+    /// Returns a pointer and the actual size of the allocated block. The pointer is suitable for
+    /// holding data described by a new layout with `layout`’s alignment and a size given by
+    /// `new_size`. To accomplish this, the allocator may extend the allocation referenced by `ptr`
+    /// to fit the new layout.
+    ///
+    /// If this returns `Ok`, then ownership of the memory block referenced by `ptr` has been
+    /// transferred to this allocator. The memory may or may not have been freed, and should be
+    /// considered unusable (unless of course it was transferred back to the caller again via the
+    /// return value of this method).
```

Comment:
I wouldn't document this limitation any further, otherwise it could be a source for bugs. If callers always treat `ptr` as unusable, there is no problem if you change the program to use `MayMove` instead of `InPlace`.

Reply:
The whole point of `InPlace` is that you have active references into the memory you are growing/shrinking, and you don't want the allocator to invalidate them.